### PR TITLE
Issue/1638 race condition in compiles on a state with set state transition

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

Part of https://github.com/inmanta/inmanta-lsm/issues/1638

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

File: src/inmanta/protocol/methods_v2.py
Comment (sanderr) [inline src/inmanta/protocol/methods_v2.py:77]: I think we can save some complexity by making the soft delete purely the responsibility of the exporter, and leave the `put_partial` endpoint untouched. But it might be good to ask the opinion of the rest of the team.

File: src/inmanta/export.py
Comment (sanderr) [inline src/inmanta/export.py:515]: *If* we decide to make this the responsibility of the exporter rather than `put_partial`, as suggested in another comment, I think we can even lift the logic to the `run` method.

File: changelogs/unreleased/1638-add-soft-delete-option.yml
Comment (sanderr) [inline changelogs/unreleased/1638-add-soft-delete-option.yml:7]: I think this bug also exists in iso6. I'm not 100% sure, did you check?

File: src/inmanta/app.py
Comment (sanderr) [inline src/inmanta/app.py:459]: I'm still thinking if I can come up with a better name. This one was always meant as a placeholder, but perhaps it's not too bad.

File: tests/agent_server/test_resource_sets.py
Comment (sanderr) [inline tests/agent_server/test_resource_sets.py:1]: Haven't reviewed these yet, depends a tiny bit on whether we consider this the responsibility of exporter or server.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/app.py
Comment (Hugo-Inmanta) [reply src/inmanta/app.py:459]: I couldn't come up with something better and it kinda grew on me.

File: changelogs/unreleased/1638-add-soft-delete-option.yml
Comment (sanderr) [inline changelogs/unreleased/1638-add-soft-delete-option.yml:6]: Right, hadn't considered that it would have to be a patch to go in iso6. Still, it's more of a bugfix addition than a real feature.

File: src/inmanta/export.py
Comment (sanderr) [inline src/inmanta/export.py:484]: I think this could be a lot simpler if you put it in `_load_resource_sets`, or am I missing something. We could even go so far as to add an attribute `self._resource_sets: set[str]` because that method already works by populating an attribute rather than returning its result.

File: src/inmanta/protocol/methods_v2.py
Comment (sanderr) [inline src/inmanta/protocol/methods_v2.py:76]: This should go now.

**Review by sanderr — CHANGES_REQUESTED** [review]

**Review by sanderr — APPROVED** [review]
Can you still request a second review?

File: changelogs/unreleased/1638-add-soft-delete-option.yml
Comment (arnaudsjs) [inline changelogs/unreleased/1638-add-soft-delete-option.yml:3]: I don't think this sentence will make a lot of sense to anyone who is no part of the core team.
```suggestion
    Add `--soft-delete` option to the exporter. This option makes sure that resource sets, specified using the `--delete-resource-set` option, are only deleted if there are no resources exported for this set.
```

File: tests/db/migration_tests/test_v202403010_to_v202403120.py
Comment (arnaudsjs) [inline tests/db/migration_tests/test_v202403010_to_v202403120.py:36]: Shouldn't we test that `soft_delete` is false on a compiles after the migration?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: docs/model_developers/resource_sets.rst
Comment (arnaudsjs) [inline docs/model_developers/resource_sets.rst:101]: I don't see why we need to the note box here. It used to attract the attention of the user to something special. This bit is not something special is it?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
